### PR TITLE
Fine tune transformer

### LINKED ISSUE #6: Fine-tune transformer notebook
We should fine-tune the transformer model in the example notebook to get sensible results.

### PR DESCRIPTION
**Pull request by geek-yang**
Fine tune transformer and generate sensible results for the example data. The following changes are made in this PR:

- A example notebook with autoencoder (a simplified transformer without complex decoder)
- Fine tune the transformer with multihead attention
- Schema images to each notebook
- Add docstrings to the source code of transformer

This PR closes #6 .

### CODE REVIEW

File: workflow/pred_temperature_LSTM.ipynb [hunk 1]
```diff
@@ -22,6 +22,16 @@
     "- Evaludate model"
```
Comment (BSchilperoort) [inline workflow/pred_temperature_LSTM.ipynb:22]: <pre class="ql-syntax" spellcheck="false">Line #2.    # check this doc if not https://era5cli.readthedocs.io/en/latest/instructions.html#register-at-copernicus-climate-data-service
</pre><p>This link is broken. The right link is now https://era5cli.readthedocs.io/en/stable/getting_started/</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/AI4S2S/cookbook/pull/7/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

File: workflow/pred_temperature_autoencoder.ipynb [hunk 2]
```diff
@@ -0,0 +1,1298 @@
+{
```
Comment (BSchilperoort) [inline workflow/pred_temperature_autoencoder.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #2.    # check this doc if not https://era5cli.readthedocs.io/en/latest/instructions.html#register-at-copernicus-climate-data-service
</pre><p>https://era5cli.readthedocs.io/en/stable/getting_started/</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/AI4S2S/cookbook/pull/7/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

Comment (BSchilperoort) [inline workflow/pred_temperature_autoencoder.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #3.    target_series_sel = target_field_resample['ts'].sel(cluster=3)
</pre><p>This cluster won't be available from the data downloaded with era5cli</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/AI4S2S/cookbook/pull/7/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='20'/>

Comment (BSchilperoort) [inline workflow/pred_temperature_autoencoder.ipynb:1]: <p>I am not sure what you mean with "but only the encoding part"</p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/AI4S2S/cookbook/pull/7/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='28'/>

Comment (BSchilperoort) [inline workflow/pred_temperature_autoencoder.ipynb:1]: <p>Loss should be low, right? So why is the validation loss so high compared to the training loss, is this overfitting?</p><p>The loss for the validation data is very high after the first few epochs.</p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/AI4S2S/cookbook/pull/7/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='34'/>

File: workflow/pred_temperature_transformer.ipynb [hunk 3]
```diff
@@ -5,7 +5,7 @@
    "cell_type": "markdown",
```
Comment (BSchilperoort) [inline workflow/pred_temperature_transformer.ipynb:5]: <p>This looks a lot better than the other notebook. Can you add some extra info with the figures? Some short analysis?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/AI4S2S/cookbook/pull/7/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='33'/>

File: workflow/pred_temperature_LSTM.ipynb (hunk 1 above)
Comment (geek-yang) [reply workflow/pred_temperature_LSTM.ipynb:22]: <p>Thanks! I will update it!</p>

File: workflow/pred_temperature_autoencoder.ipynb (hunk 2 above)
Comment (geek-yang) [reply workflow/pred_temperature_autoencoder.ipynb:1]: <p>Updated!</p>

Comment (geek-yang) [reply workflow/pred_temperature_autoencoder.ipynb:1]: <p>Yes, we need to update this when we include downloading into our workflow. I will run each notebook again to make sure that everything still works.</p>

Comment (geek-yang) [reply workflow/pred_temperature_autoencoder.ipynb:1]: <p>Thanks for noticing. In this notebook, I only use the encoder (left) part of the model, without encoding and attention for outputs (right). This is also a structure used in a famous model called BERT. But I agree that this sounds confusing. I rephrase this part and add a new figure to it. </p>

Comment (geek-yang) [reply workflow/pred_temperature_autoencoder.ipynb:1]: <p>Yes, this is terribly overfitting...Simply because the data we have is far from sufficient. I didn't try hard to improve it since we will have era5 data, which will be (slightly) longer than the testing data. I plan further tune this model later when we use the new data. For now I can simply reduce the number of epoch, to make it less overfitting.</p>

File: workflow/pred_temperature_transformer.ipynb (hunk 3 above)
Comment (geek-yang) [reply workflow/pred_temperature_transformer.ipynb:5]: <p>I would prefer to do this later when we switch to the era5 data, which is longer than the example data. For now it is hard to say, especially given that the data we have is quite short and this behavior might not be robust.</p>

**Comment by geek-yang** [comment]
Thanks for your comment @BSchilperoort! Your comments were adapted to the notebooks now. Please take a look again! Thanks a lot! 😆

File: workflow/pred_temperature_autoencoder.ipynb [hunk 4]
```diff
@@ -0,0 +1,1017 @@
...
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "#### Initialize and train model\n",
+    "Create autoencoder. Initialize model and choose loss function and optimizer. <br>\n",
+    "\n",
+    "The architecture of the autoencoder used here is shwon in the figure below. This structure is very similar to the famous language model called BERT. For more details about the full transformer network strcture, check the paper [BERT: Pre-training of Deep Bidirectional Transformers for Language Understanding](https://arxiv.org/abs/1810.04805).\n",
```
Comment (BSchilperoort) [inline workflow/pred_temperature_autoencoder.ipynb:594]:
```suggestion
    "The architecture of the autoencoder used here is shown in the figure below. This structure is very similar to the famous language model called BERT. For more details about the full transformer network structure, check the paper [BERT: Pre-training of Deep Bidirectional Transformers for Language Understanding](https://arxiv.org/abs/1810.04805).\n",
```

**Review by BSchilperoort — APPROVED** [review]
Just one small comment!

File: workflow/pred_temperature_autoencoder.ipynb (hunk 4 above)
Comment (geek-yang) [reply workflow/pred_temperature_autoencoder.ipynb:594]: Well spot!
